packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -8,13 +8,813 @@ import 'colors.dart';
+        columnWidths = <double>[
+          estimatedColumnWidths[_pickerColumnType.dayOfMonth.index],
+          estimatedColumnWidths[_pickerColumnType.month.index],
+          estimatedColumnWidths[_pickerColumnType.year.index]];
+        break;
+      case DatePickerDateOrder.ymd:
+        pickerBuilders = <Function>[_buildYearPicker, _buildMonthPicker, _buildDayPicker];
+        columnWidths = <double>[
+          estimatedColumnWidths[_pickerColumnType.year.index],
+          estimatedColumnWidths[_pickerColumnType.month.index],
+          estimatedColumnWidths[_pickerColumnType.dayOfMonth.index]];
+        break;
+      case DatePickerDateOrder.ydm:
+        pickerBuilders = <Function>[_buildYearPicker, _buildDayPicker, _buildMonthPicker];
+        columnWidths = <double>[
+          estimatedColumnWidths[_pickerColumnType.year.index],
+          estimatedColumnWidths[_pickerColumnType.dayOfMonth.index],
+          estimatedColumnWidths[_pickerColumnType.month.index]];
+        break;
+      default:
+        assert(false, 'date order is not specified');
+    }
+
+    final List<Widget> pickers = <Widget>[];
+
+    for (int i = 0; i < columnWidths.length; i++) {
+      final double offAxisFraction = (i - 1) * 0.5 * textDirectionFactor;
+
+      EdgeInsets padding = const EdgeInsets.only(right: _kDatePickerPadSize);
+      if (textDirectionFactor == -1)
+        padding = const EdgeInsets.only(left: _kDatePickerPadSize);
+
+      pickers.add(LayoutId(
+        id: i,
+        child: pickerBuilders[i](
+          offAxisFraction,
+              (Widget child) {
+            return new Container(
+              alignment: i == columnWidths.length - 1
+                  ? alignCenterLeft
+                  : alignCenterRight,
+              padding: i == 0 ? null : padding,
+              child: new Container(
+                alignment: i == 0 ? alignCenterLeft : alignCenterRight,
+                width: columnWidths[i] + _kDatePickerPadSize,
+                child: child,
+              ),
+            );
+          },
+        ),
+      ));
+    }
+
+    return new MediaQuery(
+      data: const MediaQueryData(textScaleFactor: 1.0),
+      child: new NotificationListener<ScrollEndNotification>(
+        onNotification: (ScrollEndNotification notification) {
+          // Whenever scrolling lands on an invalid entry, the picker
+          // automatically scrolls to a valid one.
+          int desiredDay = new DateTime(selectedYear, selectedMonth, selectedDay).day;
+          if (desiredDay != selectedDay) {
+            // dayController.jumpToItem() won't work here, at least for now,
+            // because jumpToItem() calls to goIdle(), which will trigger a
+            // ScrollEndNotification at the same scroll position and therefore
+            // leads to an infinite loop.
+            // Animates at super speed is used instead.
+
+            // animateToItem() is not working properly.
+            dayController.animateToItem(
```

Comment:
Just keep the future returned here and and check that the future is null or completed in the if check above.